coverage ext - config to log missing coverage

Cielquan opened the pull request:
Subject: coverage ext - config to log missing coverage


### Feature or Bugfix
<!-- please choose -->
- Feature


### Purpose
When running the `coverage` builder the output is written to files only. This feature let the output also be printed to console.

One can then also use the `coverage` builder in CI pipelines as a check with `-W` flag.


### Detail
- Added `coverage_print_missing_c_items` config var (default: `False`)
  When set to `True` undocumented c api elements are logged on `info` level.
  When a `quiet` or `warning is error` flag is set undocumented c api elements are logged on `warning` level (like `linkcheck` builder does)
- Added `coverage_print_missing_py_items` config var (default: `False`)
  When set to `True` undocumented python objects are logged on `info` level.
  When a `quiet` or `warning is error` flag is set undocumented python objects are logged on `warning` level (like `linkcheck` builder does)

### Relates
- Fixes #7758

Cielquan committed:
added config to log missing c/py coverage

tk0miya commented on sphinx/ext/coverage.py:
```diff
@@ -121,6 +124,14 @@ def write_c_coverage(self) -> None:
+                    if self.config.coverage_print_missing_c_items:
+                        if self.app.quiet or self.app.warningiserror:
+                            logger.warning(__('undocumented c api: %s [%s] in file %s'),
+                                           name, typ, filename)
```
The warning level log is not good for this case (including linkcheck builder...). With `-W` option, sphinx-build fails ASAP if any warnings found. But it would be better to show full error on CI process for me.

tk0miya replied:
Ah, I'd forgotten about `--keep-going` option. It helps `-W` on the CI server. Please forget my comment above.

Cielquan replied:
I run sphinx-build two times:
1. with `-nq` to get all warnings
2. with `-nQW` to get fail/success status.

I copied this form somewhere else. But maybe the `-keep-going` option can reduce this to one run.

tk0miya commented on sphinx/ext/coverage.py:
```diff
@@ -121,6 +124,14 @@ def write_c_coverage(self) -> None:
                 for typ, name in sorted(undoc):
                     op.write(' * %-50s [%9s]\n' % (name, typ))
+                    if self.config.coverage_print_missing_c_items:
+                        if self.app.quiet or self.app.warningiserror:
```
I think the runtime option `quiet` is stronger than a configuration. So it would be better to suppress these messages.

Cielquan replied:
I copied this behavior from `linkcheck` [here](https://github.com/sphinx-doc/sphinx/blob/3.x/sphinx/builders/linkcheck.py#L258-L262).
I wrapped the part with `if self.config.coverage_show_missing_items:` because I wanted to keep the default behavior which is nothing gets printed/logged.

How do you @tk0miya want me to change the behavior?
When should:
1. nothing be logged?
2. be logged at `info` level?
3. be logged at `warning` level?

Current behavior:
1. When `coverage_show_missing_items=False` (default and independent on other settings)
2. When `coverage_show_missing_items=True` **and neither** `-q` **nor** `-W` 
3. When `coverage_show_missing_items=True` **and** `-q` **and/or** `-W`

tk0miya replied:
I thought nothing logged is better at first. But it is better to do the same as linkcheck does. Okay, keep as is.

tk0miya commented on sphinx/ext/coverage.py:
```diff
@@ -121,6 +124,14 @@ def write_c_coverage(self) -> None:
+                        else:
+                            logger.info(red('undocumented  ') + darkgreen('c   ') +
+                                        darkblue('api       ') + '%-30s' % (name + " [%9s]" % typ) +
+                                        red(' - in file ') + filename)
```
I feel this message is too much colorized... Could you make this simple? We have to support who have poor eyesight (see #5989) in nearly future.

Cielquan replied:
Sure will try.

Cielquan replied:
![info log](https://user-images.githubusercontent.com/43916661/83355232-6da4ed80-a35e-11ea-970c-3f5737201eb8.PNG)

I removed the colors on 2nd and 3rd columns. I think this is fine?

tk0miya replied:
LGTM!

tk0miya commented on sphinx/ext/coverage.py:
```diff
@@ -239,16 +250,46 @@ def write_py_coverage(self) -> None:
                     if undoc['funcs']:
                         op.write('Functions:\n')
                         op.writelines(' * %s\n' % x for x in undoc['funcs'])
+                        if self.config.coverage_print_missing_py_items:
```
Is there any reason to separate configrations for c_items and py_items? If none, it is enough to add one option for such purpose.

tk0miya replied:
BTW, how about using a keyword "show" for this configuration instead of "print"?

Cielquan replied:
There is no special reason for the split. I will change it.

"show" sounds better.

Cielquan replied:
`coverage_show_missing_items` is the new option

tk0miya replied:
Sounds good!

Cielquan committed:
merged new config options into coverage_show_missing_items

Cielquan committed:
fixed line length

Cielquan committed:
reduced colors on info level

Cielquan committed:
fixed color imports in coverage.py

Cielquan committed:
added test for show_missing_items True

Cielquan commented:
I added 2 asserts to test_ext_coverage.test_build to check that no output is shown on console.
I added another test to check for output on console when `coverage_show_missing_items=True`

@tk0miya Can you tell me how to set cli flags for the app instance in tests?
I want to set the `-q` flag to test the warning output.

tk0miya commented:
Please set `app.quiet = True` directly before the build.

Actually, a StringIO object is assigned to `app._warning` when quiet mode. But it is already set in test script. So only we can do is setting up the `app.quiet` flag to True.
https://github.com/sphinx-doc/sphinx/blob/60f4b45bbe1da75e466042b50fcb96f437f776af/sphinx/application.py#L176-L186

Cielquan committed:
added test for show_missing_items True in quiet mode

Cielquan commented:
Thanks :)
I added another test to test the output on warning level if `coverage_show_missing_items` is True and the app set to quiet.

tk0miya reviewed (changes requested):
Sorry, I noticed document for coverage extension is not updated yet.
Could you add a document for `coverage_show_missing_items` to `doc/usage/extensions/coverage.rst`? Then I'll merge this soon.

Cielquan committed:
documented coverage_show_missing_items confval

Cielquan commented:
> Sorry, I noticed document for coverage extension is not updated yet.
> Could you add a document for `coverage_show_missing_items` to `doc/usage/extensions/coverage.rst`? Then I'll merge this soon.

At the beginning of this PR I had it in mind to change the docs also, but then I forgot eventually 😒 
Thanks for the reminder 👍

tk0miya reviewed (approved):
Good work! LGTM :-)

tk0miya commented:
Just merged. It will be released as 3.1.0 soon.
